qep-342-charts.md

```diff
@@ -0,0 +1,123 @@
+The charting capabilities will then be exposed through the layout designer
+via a new layout charts item.
+
+### Qgs{Line,Bar,Pie}Chart classes
+
+A new series of chart classes - QgsLineChart, QgsBarChart, and QgsPieChart -
+will be added to render these specific chart items. The classes would be
+parented to pre-existing QgsPlot classes and re-impelment the render( QgsRenderContext & )
+and renderContent( QgsRenderContext &, QRectF ) functions.
+
+Preparatory work will be needed to split Qgs2DPlot into a generic Qgs2DPlot
+and a specialized Qgs2DXyPlot so axis-less graphs such as QgsPieChart can
+be parented to the generic Qgs2DPlot.
+
+All three chart style implementations would rely on QgsSymbol classes to
+render its data components. E.g., for the bar chart, a QgsFillSymbol would
+be used to render the bars. This will ensure that rendered charts within
+print and atlas layouts look and feel exactly the same as other items and map
+canvases. The chart rendering will inject a couple of variables into a chart
+scope to allow for users to leverage data-defined properties.
+
+### Chart data handling: QgsPlotData
+
+A new QgsPlotData class will be added to cover generic plots/charts series
+used to render the charts. The class will also hold the categories used
+to render charts when the X axis type is set to represent categories.
+
+A QgsAbstractPlotSeries abstract class will be added, holding basic information
+such as the series name as well as styling components. A QgsXyPlotSeries
+implementation to hold the data used to render line, bar, and pie charts.
+The class will return raw values stored in a QList of QPair<double, double>.
+
+The Qgs2DPlot class’ render( QgsRenderContext & ) and renderContent( QgsRenderContext &, QRectF )
+functions will be tweaked to add a new QgsPlotData pointer. This is possible as
+the plot classes are currently marked as unstable.
+
+Finally, a new QgsVectorLayerPlotDataGenerator class will be added that takes
+care of generating plot data from vector layers. The class will iterate through
+features to prepare the plot data based on a provided set of user configured
+series. The iteration and data preparation will happen in a non-blocking
+manner off the main thread.
+
+### Layout chart item
+
+On the GUI front, this proposal aims at exposing charting through the layout
+designer. A new chart item will be added with relevant widgets to configure
+the chart’s style as well as a vector layer-driven dataset setup. 
+
+On the rendering front, when rendering a chart in “preview” mode within the
+layout designer, the dataset preparation will happen in a non-blocking way off
+the main thread, with a busy/synchronization indicator shown as a placeholder
+until the dataset is prepared and ready to be rendered. This mimics what is
+done with other items such as the map item as well as the elevation profile item.
+
+The chart item will be able to take into account an atlas context to allow
+for atlas-driven charts.
```

Comment:
What do you mean by this? Is it filtering content to objects within the report feature?

Reply:
Implementation wise, this will happen in the plot data generator / gatherer, merely a filtering against a given vector layer using atlas scope variables. It's one of the nice things it'll do , just thought it'd be great to mention it here :)